[CJ-783] iP

CJ-783 opened the pull request:
# Echo

> Echo your ideas into a list


Echo frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is,

1. download it from [here](https://github.com/CJ-783/ip/releases/tag/v0.1).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE!**

Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)



If you are a Java programmer, you can use it to practice Java too. Here's the main method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

CJ-783 committed:
Change welcome message

Welcome message contains default values (Duke)

Change the default value to "Echo", our selected name for the chatbot

Let's
* Change the filename to "Echo"
* Change the welcome banner to contain "Echo"
* Change filename to "Echo"

CJ-783 committed:
Include echo functionality

Added echo functionality to the chatbot

CJ-783 committed:
Add list and store functionality

Current functionality only contains echo.

Let's
* allow user to store their input
* return the list that the user stores

CJ-783 committed:
Created new class for Task

When a user type in an input, it will be added as a task object.
The task entered also contains many different data (e.g. description,
status etc.).

Let's create a new class called "Task" containing all the different
data and functionality.

CJ-783 committed:
Update Echo.java

With the introduction of the Task class, we will update the code in the
main class to incorporate the Task class.

Let's
* change all the task element into the task object

CJ-783 committed:
Update Task.java

Printing out Task object is manually typed out in the main file instead
of being done by the Task Object

Let's
* update Task.java to include the printing of String

CJ-783 committed:
Update Echo.java

The printing is manually typed instead of utilizing Task object

Let's
* update the main method to print from Task object itself

CJ-783 committed:
Create Deadline, Event, Todo class

Deadline, Event, Todo are tasks, but with additional specific
functionality.

Let's
* Create these additional class file to ensure that the code are more
neat.

CJ-783 committed:
Update Echo.java

Current app does not allow the different type of tasks (todo, event,
deadline).

Let's
* Add the functionality of events, todo and deadline into our chatbot

CJ-783 committed:
Add handling of error

Text entered by users are unpredictable. If there is incorrect input,
it might lead to an error output

Let's
* Handle error related to wrong input
* Handle error related to empty description

CJ-783 committed:
Update Echo.java to incorporate arraylist instead of array

Arraylist offers dynamic sizing and the ability to easily find, add,
and delete items.

Let's
* Change the type Array to Arraylist

CJ-783 committed:
Add delete functionality

To allow the users to delete tasks after they have completed it.

Let's
* add delete functionality into the chatbot

CJ-783 committed:
Add automated Text UI Testing

When we change a functionality, or add functionality, we will have to
retest all the cases

Let's
* Created an automated testing

CJ-783 committed:
Clean code for clarity and maintainability

The code contains long methods and redundant statements.

This will result in difficulty in maintaining it.

Let's
* Create a new method for welcome message
* shorten our method for updating list
* change to switch statement

CJ-783 committed:
Add writing to file functionality

The chatbot is unable to write it's Task list onto the file.

This results in the list disappearing whenever the bot reboots.

Let's update the chatbot to allow the user to save into a file.

CJ-783 committed:
Merge branch 'branch-Level-7'

CJ-783 committed:
Bugs Fix: amend the splitting of user input

Additional space added to the different information when splitting
user's input

This could cause additional bugs in the future

Let's amend the way we split our user input to fix the bug

CJ-783 committed:
Add Dates and Times functionality

When the user key in a date & time, the value is stored as a String

This reduces the functionality when we store the date and time

Let's Change the format of the stored date to be LocalDateTime

CJ-783 committed:
Bug Fix: Remove print statements

The output results in print statements

This will result in unwanted output

Let's remove all the print statement

CJ-783 committed:
Merge branch 'branch-Level-8'

CJ-783 committed:
Add additional to make the code more OOP

Majority of the functionality is in Echo.java

This makes it difficult to modify and maintain the code

Let's create additional classes to utilize the strength of OOP. Class
created will be as follows
* Parser: parse user's input
* Storage: saving and retrieving information in the file
* TaskList: information about the task list
* Ui: interaction with the user

CJ-783 committed:
Change the Basic functionality of the chatbot into OOP

The current code may result in difficulty maintaining.

This is due to the code not fully utilizing OOP standard.

Let's change the basic functionality into OOP standard before adding
the rest.

CJ-783 committed:
Add Saving task list to a text file in a more OOP way

Let's amend the Storage.java file to include the functionality of
saving it to a file

CJ-783 committed:
Add the ability to load data from file

Currently, when the user use the chatbot, the information keyed in
before the chatbot was restarted will be lost

This is will cause the user to retype all their todo list

Let's create the ability to load from the file into the chatbot

CJ-783 committed:
Remove all debugging statements

CJ-783 committed:
Merge tag 'A-MoreOOP'

CJ-783 committed:
Divide the files into its own package

The current classes are in a single file

This makes it difficult to maintain

Let's divide the classes into its own packages

CJ-783 committed:
Merge tag 'A-Packages'

CJ-783 committed:
Add gitignore file

CJ-783 committed:
Update runtest.bat file to include updated code

CJ-783 committed:
Merge branch 'add-gradle-support'

CJ-783 committed:
Add JUnit for automated testing purposes

We are using a script to test out input and output.

This will be difficult to manage in the future.

Let's add JUnit and test cases to help with our automated testing. The
test cases include the basic input for "Event" and "Todo" tasks.

CJ-783 committed:
Bug: Adding an event contains an additional output on screen

CJ-783 committed:
Bug: Fixed the inconsistency in spacing when displaying and saving
deadlines and event

CJ-783 committed:
Add DeadlineTest class

There is not class for the testing of deadline's class.

Not testing deadline class might pose issues in the future.

Let's create test cases and deadlineTest class.

CJ-783 committed:
Merge tag 'A-JUnit'

CJ-783 committed:
Add additional test cases for Task

There is currently not test cases for Task class.

There could be issue in the future relating to the Task class.

Let's create a TaskTest class for the purpose of testing the methods.

CJ-783 committed:
Merge tag 'A-JUnit'

CJ-783 committed:
Update Readme file

Current readme file contains the default name.

This might cause confusion.

Let's update the README file to contain the correct name.

CJ-783 committed:
Change the date format when printing the deadline

The current date printed shows the exact date.

This is less easier to read as compared to writing the month out
explicitly.

Let's change the date time format to print the deadline in MMM dd yyyy
format.

CJ-783 committed:
Add Javadoc to all the classes

The current document don't contain any java doc

This might be difficult others to understand the code

Let's add javadoc to all the classes

CJ-783 committed:
Add the ability to find task

The current program does not contain the find functionality.

This limits the ability for the user to check if they have added a task.

Let's add the functionality to find task.

CJ-783 committed:
Update the code base to follow the coding statndard

CJ-783 committed:
Update the code to be in accordance to our java coding standard

CJ-783 committed:
Merge branch 'branch-A-CodingStandard'

CJ-783 committed:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/echo/Echo.java
#	src/main/java/echo/exceptions/EchoIncorrectOption.java
#	src/main/java/echo/parser/Parser.java
#	src/main/java/echo/storage/Storage.java
#	src/main/java/echo/tasklist/TaskList.java
#	src/main/java/echo/tasks/Deadline.java
#	src/main/java/echo/tasks/Event.java
#	src/main/java/echo/tasks/Task.java
#	src/main/java/echo/tasks/Todo.java

CJ-783 committed:
Amend the files to fix incomplete merging of the conflicts

CJ-783 committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/echo/Echo.java
#	src/main/java/echo/parser/Parser.java
#	src/main/java/echo/tasklist/TaskList.java
#	src/main/java/echo/ui/Ui.java

CJ-783 committed:
bug fix: Encounter error when Echo file is not found

CJ-783 committed:
Bug Fix: Able to parse dates of different type from saved file

CJ-783 committed:
Update gradle to enable check-style

GaaneshT commented on README.md:
```diff
@@ -13,14 +13,16 @@ Prerequisites: JDK 17, update Intellij to the most recent version.
-   | | | | | | | |/ / _ \
-   | |_| | |_| |   <  __/
-   |____/ \__,_|_|\_\___|
+                #######  #####   #     #  #######
+                #       #     #  #     #  #     #
+                #       #        #     #  #     #
+                #####   #        #######  #     #
+                #       #        #     #  #     #
+                #       #     #  #     #  #     #
+                #######  #####   #     #  #######;
```
I like this ASCII art 😄

GaaneshT commented on src/main/java/echo/Echo.java:
```diff
@@ -0,0 +1,68 @@
+package echo;
+
+//import java.io.File;
```
I think it would be good to remove this unused import but overall it looks clean to me! :)

GaaneshT commented on src/main/java/echo/parser/Parser.java:
```diff
@@ -0,0 +1,180 @@
+    }
+
+
+    private void findBook(String userDesc) {
+        ArrayList<Task> tasksFound = this.taskList.findTask(userDesc);
+        ui.findBook(tasksFound);
+    }
```
I think a clearer name would be findTask() for this method.

GaaneshT commented on src/main/java/echo/parser/Parser.java:
```diff
@@ -0,0 +1,180 @@
+     * @param userInput The user input.
+     * @return          An integer for the Echo class to determine if the user has chosen to exit the program
+     */
+    public int getOption(String userInput) {
+        String userOption = userInput.split(" ")[0].trim().toLowerCase();
+
+        if (userOption.equals("bye")) {
+            ui.exitMessage();
+            return 0;
+        } else if (userOption.equals("list")) {
+            ui.printList(taskList);
+            return 1;
+        } else if (userOption.equals("unmark")) {
+            unMark(userInput);
+            return 1;
+        } else if (userOption.equals("mark")) {
+            mark(userInput);
+            return 1;
+        } else if (userOption.equals("find")) {
+            String userDesc = userInput.split(" ")[1].trim().toLowerCase();
+            findBook(userDesc);
+
+            return 1;
+        } else {
+            return addRemoveList(userInput);
+        }
+    }
```
hmm, why not use a switch statement here? I think it would look cleaner.

GaaneshT commented on src/main/java/echo/tasklist/TaskList.java:
Good use of javadocs here, really clean way of structuring it!

GaaneshT reviewed (commented):
LGTM!

CJ-783 committed:
Add graphic user interface functionality.

CJ-783 committed:
Merge tag 'Level-10'

CJ-783 committed:
Remove unnecessary comments

CJ-783 committed:
Merge branch 'branch-Level-10'

CJ-783 committed:
Remove class that is not in used (Duke.java)

CJ-783 committed:
Merge branch 'master' into branch-Level-10

CJ-783 committed:
Remove unnecessary methods

CJ-783 committed:
Merge branch 'branch-Level-10'

CJ-783 committed:
Add asserts functionality

There could be assumptions that the coder made that results in error.

To prevent this from happening, "assert" will be utilized throughput
the code.

CJ-783 committed:
Delete dead code

The code contains dead code in multiple lines.

This might lead to confusion in the future.

Let's delete the dead code in our files.

CJ-783 committed:
Update build gradle to include checkstyle

The checkstyle in build.gradle is not active (commented out).

This prevents us from checking if we're following the style.

Let's add in the checkstyle in the build.gradle.

CJ-783 committed:
Merge pull request #1 from CJ-783/branch-A-Assertions

Add asserts functionality

CJ-783 committed:
Merge branch 'master' of https://github.com/CJ-783/ip

* 'master' of https://github.com/CJ-783/ip:
  Add asserts functionality

CJ-783 committed:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/echo/ui/Ui.java

CJ-783 committed:
Merge pull request #2 from CJ-783/branch-A-CodeQuality

Merge A-CodeQuality branch

CJ-783 committed:
Merge branch 'master' of https://github.com/CJ-783/ip

* 'master' of https://github.com/CJ-783/ip:
  Update build gradle to include checkstyle
  Delete dead code

CJ-783 committed:
Clean up code in all the file to match Checkstyle Audit

To ensure ease of maintainability in the future and clean code.

Let's follow the CheckStyle Audit and amend accordingly

CJ-783 committed:
Merge pull request #3 from CJ-783/branch-A-CodeQuality

Clean up code in all the file to match Checkstyle Audit

CJ-783 committed:
Merge branch 'master' of https://github.com/CJ-783/ip

* 'master' of https://github.com/CJ-783/ip:
  Clean up code in all the file to match Checkstyle Audit

CJ-783 committed:
Add checking for duplicate functionality

The Task application don't have the functionality to check for
duplicates.

The user might accidentally add a duplicate task.

Let's update the application to include the functionality to detect
for duplicate tasks.

CJ-783 committed:
Bug Fix: fixing bugs related to deadline function

CJ-783 committed:
Bug fix: Deletion of task cause an error

CJ-783 committed:
Merge branch 'C-DetectDuplicates'

CJ-783 committed:
Bug fix: Deletion of an index that does not exist cause an error

CJ-783 committed:
Bug Fix: Issues with displaying date and time when adding deadline

CJ-783 committed:
Merge pull request #5 from CJ-783/bugfix/deadline-date-issue

Bug Fix: Issues with displaying date and time when adding deadline

CJ-783 committed:
Merge branch 'master' of https://github.com/CJ-783/ip

* 'master' of https://github.com/CJ-783/ip:
  Bug Fix: Issues with displaying date and time when adding deadline

CJ-783 committed:
Bug fix: test case for deadline cause an error if not handled

CJ-783 committed:
Bug fix: test cases for creation of deadline object causes error

CJ-783 committed:
Remove unnecessary comments

CJ-783 committed:
Refactor processTask method to follow guideline

CJ-783 committed:
Refactor getOption method from Parser class to improve readability

CJ-783 committed:
Refactor the method updateList

CJ-783 committed:
Merge pull request #6 from CJ-783/refactoring

Refactoring the code to be in accordance to the guideline given

CJ-783 committed:
Merge branch 'master' of https://github.com/CJ-783/ip

* 'master' of https://github.com/CJ-783/ip:
  Refactor the method updateList
  Refactor getOption method from Parser class to improve readability
  Refactor processTask method to follow guideline
  Remove unnecessary comments

CJ-783 committed:
Bug fix: Issues when loading up deadline task

CJ-783 committed:
Update docs/README doc to include deadline feature

CJ-783 committed:
Fix bug: Unable to find tasks due to capitalization

CJ-783 committed:
Update docs/README.md to include the guide for all features

CJ-783 committed:
Update docs/README.md to improve formatting

CJ-783 committed:
Update docs/README.md to update formatting for clarity

CJ-783 committed:
Update docs/README.md to improve formatting

CJ-783 committed:
Update docs/README.md update formatting for clarity

CJ-783 committed:
The JUnit test fails due to incorrect assertions in the test cases

This needs to change because the test does not correctly validate the
expected behavior, which would lead to error in testing.

Fix the assertions to match the expected output and ensure the test
cases reflect the actual behavior of the code.

This is done to ensure the correctness of the test and improve
the reliability of the automated test.

No functional changes to the main codebase, only test corrections.

CJ-783 committed:
Update methods with Javadocs

CJ-783 committed:
Improve format and clarity of docs/README

CJ-783 committed:
Add content list to README

CJ-783 committed:
Fix bug in content list display in README

CJ-783 committed:
Fix bug in content list display in README

CJ-783 committed:
Fix bug in content list display in README

CJ-783 committed:
Fix bug where content list in README doesn't show links

CJ-783 committed:
Add program title to "Echo"

CJ-783 committed:
Change Echo instance handling

CJ-783 committed:
Center-align "Send" button

CJ-783 committed:
Make "Send" button and user input textbox expand with main window

CJ-783 committed:
Make user picture circular

CJ-783 committed:
Resize user image to fit window properly

CJ-783 committed:
Improve text box design with rectangle, shadow, and padding

CJ-783 committed:
Merge branch 'branch-updating-UI'

CJ-783 committed:
Update Ui image

CJ-783 committed:
Update docs/README.md for formatting purposes

CJ-783 committed:
Update docs/README to improve formatting purposes

CJ-783 committed:
Update docs/README to improve formatting

CJ-783 committed:
Update docs/README to improve readability

CJ-783 committed:
Update docs/README to improve readability

CJ-783 committed:
Update Java documentation

CJ-783 committed:
Update docs/README to improve readability

CJ-783 committed:
Bug fix: unmark, mark, delete out of range will cause an error

CJ-783 committed:
Bug Fix: delete, unmark, mark with empty description result in error

CJ-783 committed:
Bug fix: mark, unmark, delete out of range result in error

CJ-783 committed:
Feature: close window when user enters "bye"

CJ-783 committed:
Bug fix: Issues where long strings result in overflow

CJ-783 committed:
Bug fix: adding a "/" results in wrong description

CJ-783 committed:
Update README.md
